### PR TITLE
Move the details of C99 identifier mangling out of `ProjectModel`

### PR DESCRIPTION
**Pull request by abertelrud**
Move the details of C99 identifier mangling out of `ProjectModel` and down into `Basic`.  It looks as if at some point this was meant to be able to fail because it might remove all the characters and end up with an empty string, but now that it replaces characters it doesn't seem possible for it to end up with an empty string assuming it was given a non-empty string.

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
I wonder if this should be in utilities

File: Sources/Basic/StringConversions.swift [hunk 1]
```diff
@@ -73,4 +73,212 @@ public extension String {
     public mutating func shellEscape() {
         self = shellEscaped()
     }
+    
+    /// Creates a string that contains only value C99 characters.
+    /// FIXME: We will likely want a better name...
+    public func mangledForC99() -> String {
```
Comment (ddunbar) [inline Sources/Basic/StringConversions.swift]: I think this should probably be named to follow "C99ExtendedIdentifier" which is the technical standard it follows.

**Comment by ddunbar** [comment]
The previous version would throw if the result was an empty string, because it doesn't mangle the first character (IIRC)?

**Comment by abertelrud** [comment]
@aciidb0mb3r Yes, `Utilities` would probably be better.  I put it together with the shell escaping code, but `Utilities` is probably a better home for both of them.

**Comment by abertelrud** [comment]
@ddunbar I addressed this in the commit message actually.  It looked like the code had originally been written to potentially remove invalid characters, which could lead to an empty string even when the input wasn't empty.  But at least in the sate of the code as I found it, that couldn't happen, since it always replaced unsupported scalars with `_`.  Also, the error being thrown was project-model related and not about C99 identifier mangling in general.  We can make it throw again, but I think the better thing to do is to either: a) precondition that the string isn't empty, or b) make sure the output string is never empty.  If the point here is to take an arbitrary input string and make it valid C99, it doesn't seem that throwing is a good solution.

**Comment by abertelrud** [comment]
And as I mentioned, I don't think it would actually ever throw, because it no longer makes the string shorter.

**Comment by ddunbar** [comment]
Ah, I see your point. I do remember it doing that at one point, it must have changed (for all I know it might have been me that changed it at some point).

**Comment by abertelrud** [comment]
@aciidb0mb3r @ddunbar  Do you think this should be in `Utilities` instead?  If so, shouldn't the shell-escaping code also move down?

**Comment by aciidb0mb3r** [comment]
I do think this should be in Utilities because this is not general enough for basic. I am not sure about shell escaping, It feels like a Basic thing but I wouldn't mind it being in utilities.

**Comment by ddunbar** [comment]
I agree Utilities seems better although I would feel much happier about that if we had a good name for Utilities.

**Comment by abertelrud** [comment]
Ok.  I will move this to `Utilities` and possibly move `ShellEscaping` in a separate PR.

**Comment by abertelrud** [comment]
This got stalled, going to make the discussed changes, test, and merge this now.  I am not going to rename `Utilities` at the moment, would be a different PR.

**Comment by abertelrud** [comment]
@swift-ci please test

**Comment by abertelrud** [comment]
(that test was requested to see if the CI incremental-build changes are working — I know this still needs to be rebased)

**Comment by abertelrud** [comment]
@swift-ci please test linux

**Comment by abertelrud** [comment]
@swift-ci please test linux platform

**Comment by abertelrud** [comment]
@swift-ci please test

**Comment by abertelrud** [comment]
@swift-ci please test

**Comment by abertelrud** [comment]
I've moved it to `Utilities` now.  I'll also add a performance test, but wanted to update the PR and kick off the correctness testing.

**Comment by abertelrud** [comment]
Hmm, those failures suggest that not all the diffs got pushed to the PR (I specifically fixed the problem it's erroring out about).  Investigating.

**Comment by abertelrud** [comment]
Turns out I hadn't pushed the latest changes.  Testing again.

**Comment by abertelrud** [comment]
@swift-ci please test

**Comment by abertelrud** [comment]
The Linux failure is puzzling, since apparently the build and the tests all succeeded.  The failure seems to be in the automation that happens after that.

**Comment by abertelrud** [comment]
I'll do some more local testing before merging.

File: Tests/UtilityTests/StringManglingTests.swift [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        
+        // Invalid non-leading characters.
+        XCTAssertEqual("_-".mangledToC99ExtendedIdentifier(), "__")
+        XCTAssertEqual("foo-bar".mangledToC99ExtendedIdentifier(), "foo_bar")
+        
+        // Invalid leading characters.
+        XCTAssertEqual("1".mangledToC99ExtendedIdentifier(), "_")
+        XCTAssertEqual("1foo".mangledToC99ExtendedIdentifier(), "_foo")
+        
+        // FIXME: There are lots more interesting test cases to add here.
```
Comment (aciidb0mb3r) [inline Tests/UtilityTests/StringManglingTests.swift]: Finally a chance to use emoji in swiftpm source code 😀

**Pull request merged**